JBPM-5246: Case Actions. DynamicSubprocess and pagination added

nmirasch opened the pull request:
Added Actions Pagination and add Dynamic Subprocess available action
@cristianonicolai can you take a look?

cristianonicolai commented on jbpm-wb-case-mgmt/jbpm-wb-case-mgmt-client/src/main/java/org/jbpm/workbench/cm/client/actions/CaseActionsPresenter.java:
```diff
@@ -96,10 +154,18 @@ protected void refreshData(final boolean refreshAvailableActions) {
         caseService.call((Actions actions) -> {
             if (refreshAvailableActions) {
                 List<CaseActionSummary> availableActions = new ArrayList<>();
+                //availableActions.add(CaseActionSummary.builder()
```
Could this be removed?

cristianonicolai commented on jbpm-wb-case-mgmt/jbpm-wb-case-mgmt-client/src/main/java/org/jbpm/workbench/cm/client/actions/NewActionViewImpl.java:
```diff
@@ -144,39 +181,75 @@ private boolean validateForm() {
-            assignationHelp.setTextContent(translationService.format(PLEASE_INTRO_USER_OR_GROUP_TO_ASSIGN_NEW_ACTION));
-            validForm = false;
+        switch (caseActionType) {
+            case ADD_DYNAMIC_TASK: {
```
If we don't support dynamic task via the UI I guess this could be removed.

cristianonicolai commented on jbpm-wb-case-mgmt/jbpm-wb-case-mgmt-client/src/main/java/org/jbpm/workbench/cm/client/actions/NewActionViewImpl.html:
```diff
@@ -38,6 +38,23 @@ <h4 class="modal-title" data-i18n-key="NewAction"></h4>
+                        </div>
+
+                        <div class="form-group" data-field="action-case-definitions-group">
+                            <label class="control-label col-md-3" data-i18n-key="ProcessId"></label>
```
Please add a required indicator to this field. Also perhaps name it as _Process Name_ instead.

cristianonicolai commented on jbpm-wb-case-mgmt/jbpm-wb-case-mgmt-client/src/main/java/org/jbpm/workbench/cm/client/actions/CaseActionsPresenter.java:
```diff
@@ -64,26 +70,78 @@ protected void clearCaseInstance() {
+                    Collections.sort(definitionNames);
+                    newActionView.setCaseDefinitions(definitionNames);
+                }
+        ).getCaseDefinitions();
```
I think this call is incorrect, we should get all the processes that are not cases, which is returned by _CaseServicesClient.findProcessesByContainerId_.

cristianonicolai commented on jbpm-wb-case-mgmt/jbpm-wb-case-mgmt-client/src/main/java/org/jbpm/workbench/cm/client/actions/CaseActionsListViewImpl.java:
```diff
@@ -38,6 +42,7 @@
 @Dependent
 @Templated
 public class CaseActionsListViewImpl extends AbstractView<CaseActionsPresenter> implements CaseActionsPresenter.CaseActionsListView {
+    private int PAGE_SIZE = 4;
```
Something tricky here, the Available actions column should be taller to match In Progress and Completed columns. One way to fix it could be adding some extra info for each line in column, that could be the stage as @bsremac suggested. I can think of three status here:
- Dynamic (For user task and process task)
- Stage name (For ad-hoc tasks from a specific case stage)
- Case level (For ad-hoc tasks from the case instance level)
wdyt?

bsremac replied:
This is definitely a good place to add clarification for the case worker about where the Ad-hoc actions are located in the model.
The same three statuses come to my mind too.
My two pennyworth when it comes to status names:
- Dynamic (would also drop "Add new dynamic" from the user/process task names)
- Available in: Stage name
- Available in: Case

nmirasch replied:
@bsremac Sorry I saw this comment after updating the pr. 
I'm going to do those changes and refresh the PR. Thanks!

nmirasch commented:
Removed all referred to generic Dynamic task and changed the service to fill the available subprocess in the dynamic suprocess addition.
@cristianonicolai Could you take another look? thanks!!!

nmirasch commented:
@cristianonicolai @bsremac changes done. Ready for review :)

bsremac commented on jbpm-wb-case-mgmt/jbpm-wb-case-mgmt-api/src/main/java/org/jbpm/workbench/cm/service/CaseManagementService.java:
```diff
@@ -69,8 +70,18 @@
 
+    void addDynamicTask(String containerId, String caseId, String nodeType, String name, Map<String, Object> data);
+
+    void addDynamicTaskToStage(String containerId, String caseId, String stageId, String nodeType, String name, Map<String, Object> data);
```
Just for consistency reasons, do we want to keep methods which handle dynamic service tasks in the interface while completely removing dynamic service task from CaseActionType enum?

bsremac commented on jbpm-wb-case-mgmt/jbpm-wb-case-mgmt-api/src/main/java/org/jbpm/workbench/cm/model/ProcessDefinitionSummary.java:
```diff
@@ -0,0 +1,153 @@
+
+@Bindable
+@Portable
+public class ProcessDefinitionSummary implements Comparable<ProcessDefinitionSummary> {
```
I don't see where in the code the implementation of Comparable interface pays off.

nmirasch commented:
@cristianonicolai @bsremac  removed the addDynamicTask and the comparator.
